src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,384 @@
+        private readonly IProvenBlockHeaderRepository provenBlockHeaderRepository;
+
+        /// <summary>Lock object to protect access to <see cref="ProvenBlockHeader"/>.</summary>
+        private readonly AsyncLock lockobj;
+
+        /// <summary>Queue which contains blocks that should be saved to the database.</summary>
+        private readonly AsyncQueue<KeyValuePair<HashHeightPair, ProvenBlockHeader>> provenBlockHeaderQueue;
+
+        /// <summary>Batch of <see cref="ProvenBlockHeader"/> items which will be saved to the database.</summary>
+        /// <remarks>Write access should be protected by <see cref="lockobj"/>.</remarks>
+        private readonly ConcurrentDictionary<HashHeightPair, ProvenBlockHeader> batch;
+
+        /// <summary>Task that runs <see cref="DequeueContinuouslyAsync"/>.</summary>
+        private Task dequeueLoopTask;
+
+        /// <summary>Maximum interval between saving batches.</summary>
+        /// <remarks>Interval value is a prime number that wasn't used as an interval in any other component. That prevents having CPU consumption spikes.</remarks>
+        private const int BatchMaxSaveIntervalSeconds = 47;
+
+        /// <summary>Maximum number of bytes the batch can hold until the downloaded blocks are stored to the disk.</summary>
+        internal const int BatchThresholdSizeBytes = 5 * 1000 * 1000;
+
+        /// <summary>The current batch size in bytes.</summary>
+        private long currentBatchSizeBytes;
+
+        /// <summary>Performance counter to measure performance of the save and query operation.</summary>
+        private readonly BackendPerformanceCounter performanceCounter;
+        private BackendPerformanceSnapshot latestPerformanceSnapShot;
+
+        /// <summary>Current block tip hash and height that the <see cref= "ProvenBlockHeader"/> belongs to.</summary>
+        public HashHeightPair TipHashHeight { get; private set; }
+
+        /// <summary>
+        /// Initializes a new instance of the object.
+        /// </summary>
+        /// <param name="network">Specification of the network the node runs on - RegTest/TestNet/MainNet.</param>
+        /// <param name="chain">Thread safe class representing a chain of headers from genesis</param>
+        /// <param name="dateTimeProvider">Provider of time functions.</param>
+        /// <param name="loggerFactory">Factory to create a logger for this type.</param>
+        /// <param name="provenBlockHeaderRepository">Persistent interface of the <see cref="ProvenBlockHeader"></see> DBreeze repository.</param>
+        /// <param name="nodeLifetime">Allows consumers to perform clean-up during a graceful shutdown.</param>
+        /// <param name="chainState">Chain state holds various information related to the status of the chain and its validation.</param>
+        /// <param name="nodeStats">Registers an action used to append node stats when collected.</param>
+        public ProvenBlockHeaderStore(
```

Comment:
`ProvenBlockHeaderStore` should only be a cache layer for proven header repository. 

There is no need to batch because we will save changes only once a minute and on shutdown (see how ChainRepository is being saved).

I think if this component has an instance of LRU memory cache and async loop that persists proven headers once a minute and cleans them from memory it will be enough

And also it should provide access to underlying repository the same way BSQ does it for BlockRepository

Reply:
Ok, the code was changed based on our conversations and the spec you detailed in this PR.